### DISCUSSION TITLE
Activate new block after some time

### DISCUSSION THREAD

**Discussion by fervaldies** [idaholab/moose#27852] [Q&A General]
### Check these boxes if you have followed the posting rules.

- [X] Q&A General is the most appropriate section for my question
- [X] I have consulted the posting Guidelines on the Discussions front page
- [X] I have searched the Discussions forum and my question has not been asked before
- [X] I have searched the MOOSE website and the documentation does not answer my question
- [X] I have formatted my post following the posting guidelines (screenshots as a last resort, triple back quotes around pasted text)

### Question

Hello, I am currently running a solidification model (in 2D), and what I want to do it to initialize a block with grains with the PolycrystalVoronoi and then (after let's say 10 seconds) add a new block on top as liquid and let the grains from that first block solidify into that new block. Then again another block on top of the previous and so on. Is there any way of doing this?

**Comment by GiudGiud** [comment]
Hello

You have a few options here:

- either create the mesh with every region there from the beginning, but use user objects to move regions to a block with equations active on it (and use controls to turn on those equations if the set of equations changes)
- use restart simulation on every change of the mesh. Each restart is initialized over the part of the mesh they share with the previous simulation with the previous solution, and with an initial condition for the new part of the domain
- create new user objects that create new elements in the mesh. This has not been done much in MOOSE yet but it is possible. You can use Controls to execute these new user objects exactly when desired

Guillaume

**Reply by fervaldies** [reply]
Thanks! And just a quick question regarding the BCs. I want to use the FunctionDirichletBC as follows:
```
  [./all]
    type = FunctionDirichletBC
    variable = T
    boundary = '0 5'
    function = bc_func
  [../]
```
with 
```
  [./bc_func]
    type = ParsedFunction
    expression = 'if(t<5,nothing,500)'
    #symbol_names = 'alpha'
    #symbol_values = '16'
  [../]
```
And I would like to not have any BC before t<5, but I do not know what to add in "nothing" to do that. If I write just 0, the boundary will go to zero during that time, but I just want the boundary to not have anything for that time and let the temperature evolve inside the mesh.

**Reply by GiudGiud** [reply]
hello

to turn on/off a boundary condition, you ll want to look at the FunctionEnableControl instead
Controls set the 'active' status of objects, and let you turn on/off equations, materials etc

**Comment by PEI0214** [comment]
Hello, I had a similar problem. Have you solved the problem?

**Reply by GiudGiud** [reply]
The user object approach should work and was further improved lately. 

Declare the future blocks in the Mesh block with the add_subdomaon_names parameter

And use the CoupledVarThresholdSubdomainModifier to move elements the new blocks when needed

**Reply by PEI0214** [reply]
Hi, I have two blocks during the simulation, both generating deformation and stress, and I want to initialize one of the blocks at some point (stress and deformation all removed).Is there any way of doing this?

**Reply by GiudGiud** [reply]
can you use the Controls system?
I am thinking you are trying to have this other block:
- exist in the mesh from the beginning
- be initialized  (or re-initialized?) at some point in the simulation
- inactive in the equations until that point

is that true?

**Reply by PEI0214** [reply]
Yes, I'm trying to simulate the filling process. When using CoupledVarThresholdElementSubdomainModifier, unfilled blocks are also deformed, as shown below. What I want to do is that only the filled completed block deforms and the rest remains unchanged (zero stress and deformation). Is there a way to do that?
![image](https://github.com/user-attachments/assets/e16e0e3a-7c2d-4c47-bb9c-16b859a89d48)

**Reply by GiudGiud** [reply]
To do that, you still use the subdomain modifier, but you make sure the equations do not cover the inactive zones. 
So in the definition of the kernels (or if you are using the Physics, of the physics), specify `block = blocks_initially_active`
then the subdomain modifier will move elements from the inactive blocks to the active  ones

**Reply by PEI0214** [reply]
Thank you for your answer. I try to simulate the fill process by subdomain modifier, and the input file is Test1.i. The problem of computation is not convergent.I then tried Test2.i as an input file. The result is shown below.  The unfilled blocks are also deformed. But that's not what I want, what I want to do is that only the filled completed block deforms and the rest remains unchanged (zero stress and deformation).  Is there any way of doing this? Thank you for your help. The input file is as follows:

- Test1.i 

~~~
[Problem]
  kernel_coverage_check = false
  material_coverage_check = false
[]

Box1_inactive_id = '4'
Box2_inactive_id = '5'
Box3_inactive_id = '6'
inactive_domain_block_ids = '${Box1_inactive_id} ${Box2_inactive_id} ${Box3_inactive_id}'

[Mesh]
  [BaseMesh]
    type = GeneratedMeshGenerator
    dim = 3
    nx = 10
    ny = 10
    nz = 30
    xmin = -0.5
    xmax = +0.5
    ymin = -0.5
    ymax = +0.5
    zmin = 0
    zmax = +3
  []
  [Box1]
    type = SubdomainBoundingBoxGenerator
    input = 'BaseMesh'
    block_id = 1
    location = INSIDE
    bottom_left = '-1 -1 0'
    top_right = '+1 +1 +3'
  []
  [Box2]
    type = SubdomainBoundingBoxGenerator
    input = 'Box1'
    block_id = 2
    location = INSIDE
    bottom_left = '-1 -1 +1'
    top_right = '+1 +1 +3'
  []
  [Box3]
    type = SubdomainBoundingBoxGenerator
    input = 'Box2'
    block_id = 3
    location = INSIDE
    bottom_left = '-1 -1 +2'
    top_right = '+1 +1 +3'
  []

  add_subdomain_ids = ${inactive_domain_block_ids}
[]

# active_domain_block_ids = '1 2 3'

[GlobalParams]
  displacements = 'disp_x disp_y disp_z'
[]

# ===== Tensor mechanics for all active domain blocks =====
[Physics/SolidMechanics/QuasiStatic]
  [all]
    add_variables = true
    strain = FINITE
    block = '${inactive_domain_block_ids}'
  []
[]

[Variables]
  [disp_x]
  []
  [disp_y]
  []
  [disp_z]
  []
[]

# ===== Gravity =====
[Kernels]
  [Gravity1]
    type = Gravity
    block = '${inactive_domain_block_ids}'
    use_displaced_mesh = false
    variable = disp_z
    value = -9.81
  []
[]

# ===== Boundary Conditions =====
[BCs]
  [archor_x]
    type = DirichletBC
    boundary = 'back'
    variable = disp_x
    value = 0
  []

  [archor_y]
    type = DirichletBC
    boundary = 'back'
    variable = disp_y
    value = 0
  []

  [archor_z]
    type = DirichletBC
    boundary = 'back'
    variable = disp_z
    value = 0
  []
[]

# ===== Materials (linear-elastic to keep it simple) =====
[Materials]
  [elasticity_tensor]
    type = ComputeIsotropicElasticityTensor
    block = '${inactive_domain_block_ids}'
    youngs_modulus = 50E6 # 50 MPa
    poissons_ratio = 0.3
  []

  [stress_base]
    type = ComputeFiniteStrainElasticStress
    block = '${inactive_domain_block_ids}'
  []

  [density]
    type = GenericConstantMaterial
    prop_names = density
    prop_values = 2150
    block = '${inactive_domain_block_ids}'
  []
[]

[UserObjects]
  [Box_SubdomainModifier]
    type = TimedSubdomainModifier
    times = '2 4 6'
    blocks_from = '1 2 3'
    blocks_to = '4 5 6'
    execute_on = 'INITIAL TIMESTEP_BEGIN'
  []
[]

# ===== Executioner =====
[Executioner]
  type = Transient

  start_time = 0
  end_time = 10
  dt = 2

  solve_type = 'PJFNK'
  petsc_options = '-snes_converged_reason'
  petsc_options_iname = '-pc_type -pc_factor_mat_solver_package'
  petsc_options_value = ' lu       mumps'

  nl_abs_tol = 1E-3
  nl_max_its = 400

  l_tol = 1E-3
  l_max_its = 200
[]

[Outputs]
  exodus = true
[]
~~~

- Test2.i

~~~
[Problem]
  kernel_coverage_check = false
  material_coverage_check = false
[]

Box1_inactive_id = '4'
Box2_inactive_id = '5'
Box3_inactive_id = '6'
inactive_domain_block_ids = '${Box1_inactive_id} ${Box2_inactive_id} ${Box3_inactive_id}'

[Mesh]
  [BaseMesh]
    type = GeneratedMeshGenerator
    dim = 3
    nx = 10
    ny = 10
    nz = 30
    xmin = -0.5
    xmax = +0.5
    ymin = -0.5
    ymax = +0.5
    zmin = 0
    zmax = +3
  []
  [Box1]
    type = SubdomainBoundingBoxGenerator
    input = 'BaseMesh'
    block_id = 1
    location = INSIDE
    bottom_left = '-1 -1 0'
    top_right = '+1 +1 +3'
  []
  [Box2]
    type = SubdomainBoundingBoxGenerator
    input = 'Box1'
    block_id = 2
    location = INSIDE
    bottom_left = '-1 -1 +1'
    top_right = '+1 +1 +3'
  []
  [Box3]
    type = SubdomainBoundingBoxGenerator
    input = 'Box2'
    block_id = 3
    location = INSIDE
    bottom_left = '-1 -1 +2'
    top_right = '+1 +1 +3'
  []

  add_subdomain_ids = ${inactive_domain_block_ids}
[]

active_domain_block_ids = '1 2 3'

[GlobalParams]
  displacements = 'disp_x disp_y disp_z'
[]

# ===== Tensor mechanics for all active domain blocks =====
[Physics/SolidMechanics/QuasiStatic]
  [all]
    add_variables = true
    strain = FINITE
    block = '${inactive_domain_block_ids} ${active_domain_block_ids}'
  []
[]

[Variables]
  [disp_x]
  []
  [disp_y]
  []
  [disp_z]
  []
[]

# ===== Gravity =====
[Kernels]
  [Gravity]
    type = Gravity
    block = '${inactive_domain_block_ids}'
    use_displaced_mesh = false
    variable = disp_z
    value = -9.81
  []
  [Gravity1]
    type = Gravity
    block = '${active_domain_block_ids}'
    use_displaced_mesh = false
    variable = disp_z
    value = 0
  []
[]

# ===== Boundary Conditions =====
[BCs]
  [archor_x]
    type = DirichletBC
    boundary = 'back'
    variable = disp_x
    value = 0
  []

  [archor_y]
    type = DirichletBC
    boundary = 'back'
    variable = disp_y
    value = 0
  []

  [archor_z]
    type = DirichletBC
    boundary = 'back'
    variable = disp_z
    value = 0
  []
[]

# ===== Materials (linear-elastic to keep it simple) =====
[Materials]

  [elasticity_tensor]
    type = ComputeIsotropicElasticityTensor
    block = '${inactive_domain_block_ids}'
    youngs_modulus = 50E6 # 50 MPa
    poissons_ratio = 0.3
  []
  [elasticity_tensor1]
    type = ComputeIsotropicElasticityTensor
    block = '${active_domain_block_ids}'
    youngs_modulus = 50E6 # 50 MPa
    poissons_ratio = 0.3
  []

  [stress_base]
    type = ComputeFiniteStrainElasticStress
    block = '${inactive_domain_block_ids}'
  []
  [stress_base1]
    type = ComputeFiniteStrainElasticStress
    block = '${active_domain_block_ids}'
  []

  [density]
    type = GenericConstantMaterial
    prop_names = density
    prop_values = 2150
    block = '${inactive_domain_block_ids}'
  []
  [density1]
    type = GenericConstantMaterial
    prop_names = density
    prop_values = 0
    block = '${active_domain_block_ids}'
  []
[]

[UserObjects]
  [Box_SubdomainModifier]
    type = TimedSubdomainModifier
    times = '2 4 6'
    blocks_from = '1 2 3'
    blocks_to = '4 5 6'
    execute_on = 'INITIAL TIMESTEP_BEGIN'
  []
[]

# ===== Executioner =====
[Executioner]
  type = Transient

  start_time = 0
  end_time = 10
  dt = 2

  solve_type = 'PJFNK'
  petsc_options = '-snes_converged_reason'
  petsc_options_iname = '-pc_type -pc_factor_mat_solver_package'
  petsc_options_value = ' lu       mumps'

  nl_abs_tol = 1E-3
  nl_max_its = 400

  l_tol = 1E-3
  l_max_its = 200
[]

[Outputs]
  exodus = true
[]
~~~
Test2 results:
![a5c938170d1a6f0ac2ce2dae9db8dd04](https://github.com/user-attachments/assets/1b8099ff-1cdb-43a9-adbe-67c3eea3a8d2)

**Reply by GiudGiud** [reply]
> But that's not what I want, what I want to do is that only the filled completed block deforms and the rest remains unchanged (zero stress and deformation).

This is about block restricting displacements. 
in your input, displacements are defined everywhere
```
[Variables]
  [disp_x]
  []
  [disp_y]
  []
  [disp_z]
  []
[]
```
but you only want to solve for their changes on part of the domain.
You should add 
```
block =
```
to each displacement to make sure they are only active where you want them to be

**Reply by PEI0214** [reply]
Thank you very much. I modified the input file according to your suggestion, and the following error occurred:
`*** ERROR ***
/home/pdd/projects1/moose/modules/solid_mechanics/problem/0730/2/test1.i:62: missing required parameter 'Physics/SolidMechanics/QuasiStatic/displacements'
    Doc String: "The nonlinear displacement variables for the problem"`
And there's a strange phenomenon. When a block is transferred from a subdomain with a gravity of 0 to a subdomain with a gravity of -9.81, its deformation continues to increase with time after the transfer is over. What causes this?

**Reply by GiudGiud** [reply]
You can solve that error by providing the required parameter to the Physics

If the elements are still deforming then there must still be an equation for displacements on that block.
Can you paste here your variables block?

**Reply by PEI0214** [reply]
The variable block is set as follows:
~~~
[Variables]
  [disp_x]
    block = '${inactive_domain_block_ids}'
  []
  [disp_y]
    block = '${inactive_domain_block_ids}'
  []
  [disp_z]
    block = '${inactive_domain_block_ids}'
  []
[]
~~~
The displacements provided in Physics still report errors.
~~~
[Physics/SolidMechanics/QuasiStatic]
  [all]
    add_variables = true
    strain = FINITE
    displacements = 'disp_x disp_y disp_z'
    block = '${inactive_domain_block_ids}'
  []
[]
~~~

**Reply by jmeier** [reply]
Dear @PEI0214, starting with the merge of https://github.com/idaholab/moose/pull/28308, Moose will enforce the block-restriction of the variables added by [Physics/SolidMechanics/QuasiStatic]. Please update our Moose (if needed) and check.
